lib/insights/api/common/audit_log/middleware.rb

```diff
@@ -0,0 +1,92 @@
+module Insights::API::Common
+  class AuditLog
+    class Middleware
+      attr_reader :logger, :evidence, :request, :status
+
+      def initialize(app)
+        @app = app
+        @logger = AuditLog.logger
+        @subscribers = []
+        @evidence = {}
+      end
+
+      def call(env)
+        subscribe
+        @request = ActionDispatch::Request.new(env)
+        @app.call(env).tap do |status, _headers, _body|
+          @status = status
+          response_finished
+        end
+      ensure
+        unsubscribe
+      end
+
+      private
+
+      def response_finished
+        payload = {
+          :controller => evidence[:controller],
+          :remote_ip  => request.remote_ip,
+          :message    => generate_message
+        }
+        log(payload)
+      end
+
+      def generate_message
+        status_label = Rack::Utils::HTTP_STATUS_CODES[status]
+        msg = "#{request.method} #{request.original_fullpath} -> #{status} #{status_label}"
+        if evidence[:unpermitted_parameters]
+          msg += "; unpermitted params #{fmt_params(evidence[:unpermitted_parameters])}"
+        end
+        if evidence[:halted_callback].present?
+          msg += "; filter chain halted by :#{evidence[:halted_callback]}"
+        end
+        msg
+      end
```

Comment:
Are the headers logged anywhere? It may be good to include, but it may also present a security risk to include the X_RH_IDENTITY header.

Reply:
Inspecting `X_RH_IDENTITY` is a good idea, and can be done here. That's mainly the reason why this is a middleware (to access the request and headers).

Reply:
The headers may also include `X_RH_REQUEST_ID` (or something similar), which can be important to tracing uploads, etc. across services.

Reply:
Awesome! I guess we can incorporate that in the second iteration.